### PR TITLE
Add some new useful iterators

### PR DESCRIPTION
**Pull request by SilensAngelusNex**

### CODE REVIEW

File: Project1/BitsetIterator.h [hunk 1]
```diff
@@ -76,79 +76,76 @@ class set_bits_iterator {
...
+private:
+	T& _container;
+	size_t _index;
 public:
-	const_bits_iterator(const std::bitset<size>& b) :
-		_bits(b),
-		_index(size)
+	index_iterator(T& container) :
+		_container(container),
+		_index(_container.size())
```
Comment (dwatchguard) [inline Project1/BitsetIterator.h:87]: Is it possible to give a type with no size method? What happens if we do, or does this ensure that it has a size?

File: Project1/BitsetIterator.h [hunk 2]
```diff
@@ -76,79 +76,76 @@ class set_bits_iterator {
...
+	index_iterator operator--(int) {
+		index_iterator result(*this);
 		--(*this);
 		return result;
 	}
 
-	bool operator*() const {
-		return _bits[_index];
+	auto operator*() {
+		return _container[_index];
```
Comment (dwatchguard) [inline Project1/BitsetIterator.h]: Again, can we give this something with no [] operation? I honestly don't know.

File: Project1/IterableBitset.h [hunk 3]
```diff
@@ -7,61 +7,75 @@
...
 	return result;
 }
 
-template<int size>
+template<size_t n>
+inline std::string vectorToString(std::vector<size_t> init) {
+	Expects(std::find_if(init.begin(), init.end(), [](size_t i) {return i > n; }) == init.end());
+
+	std::string result = std::string(n, '0');
+	for (int i : init) {
```
Comment (dwatchguard) [inline Project1/IterableBitset.h]: should this be size_t?

**Review by dwatchguard — COMMENTED** [review]
I hate data structures, but I think this all works. I really only saw one thing I think should be fixed

File: Project1/BitsetIterator.h (hunk 1 above)
Comment (SilensAngelusNex) [reply Project1/BitsetIterator.h:87]: If your type T has no member size, the resulting template type doesn't have any of the methods that would call size().
This is called [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae).

File: Project1/BitsetIterator.h (hunk 2 above)
Comment (SilensAngelusNex) [reply Project1/BitsetIterator.h]: Same answer, you won't be able to dereference it if you do.

File: Project1/BitsetIterator.h (hunk 1 above)
Comment (dwatchguard) [reply Project1/BitsetIterator.h:87]: Ok I thought it was fine, but I thought I should ask

File: Project1/IterableBitset.h (hunk 3 above)
Comment (SilensAngelusNex) [reply Project1/IterableBitset.h]: Yuuuup

**Pull request merged**